### PR TITLE
DPCF support

### PR DESCRIPTION
**Pull request by pixelflinger**

### CODE REVIEW

File: RELEASE_NOTES.md [hunk 1]
```diff
@@ -6,6 +6,7 @@ A new header is inserted each time a *tag* is created.
 ## v1.15.0 (currently main branch)
 
 - engine: Fix spotlights normal bias calculation [⚠️ **Material breakage**].
+- engine: add support for DPCF (PCF shadows with contact hardening)
```
Comment (bejado) [inline RELEASE_NOTES.md:9]: Please rebase on main and make sure this goes under the `v1.15.1` header.

File: shaders/src/shadowing.fs [hunk 2]
```diff
@@ -31,11 +44,155 @@ float ShadowSample_PCF(const mediump sampler2DArray shadowMap,
...
+ * DPCF, PCF with contact hardenning simulation.
+ * see "Shadow of Cold War", A scalable approach to shadowing -- by Kevin Myers
+ */
+float ShadowSample_DPCF(const bool DIRECTIONAL,
+        const mediump sampler2DArray map, const uint layer, const uint index, const highp vec4 shadowPosition) {
+    highp vec3 position = shadowPosition.xyz * (1.0 / shadowPosition.w);
+    highp vec2 size = vec2(textureSize(map, 0));
+    highp vec2 texelSize = vec2(1.0) / size;
+
+    // We need to use the shadow receiver plane depth bias to combat shadow acnee due to the
```
Comment (bejado) [inline shaders/src/shadowing.fs:99]:
```suggestion
    // We need to use the shadow receiver plane depth bias to combat shadow acne due to the
```

File: shaders/src/shadowing.fs [hunk 3]
```diff
@@ -31,11 +44,155 @@ float ShadowSample_PCF(const mediump sampler2DArray shadowMap,
...
+    float occludedCount = 0.0f;
+    float z_occSum = 0.0f;
+    for (uint i = 0u; i < DPCF_SHADOW_TAP_COUNT; i++) {
+        highp vec2 duv = R * poissonDisk[i] * (texelSize * penumbra);
+        float z_occ = textureLod(map, vec3(position.xy + duv, layer), 0.0).r;
+
+        // note: z_occ and z_rec are not necessarily linear here, comparing them is always okay for
+        // the regular PCF, but the "distance" is meaningless unless they are actually linear
+        // (e.g.: for the directional light).
+        // Either way, if we assume that all the samples are close to eachother we can take their
```
Comment (bejado) [inline shaders/src/shadowing.fs:134]:
```suggestion
        // Either way, if we assume that all the samples are close to each other we can take their
```

File: shaders/src/shadowing.fs [hunk 4]
```diff
@@ -209,13 +370,26 @@ float shadow(const bool DIRECTIONAL,
...
     }
 
-    // note: shadowPosition.z is in linear light-space normalized to [0, 1]
-    //  see: ShadowMap::computeVsmLightSpaceMatrix() in ShadowMap.cpp
-    //  see: computeLightSpacePosition() in common_shadowing.fs
-    highp vec3 position = vec3(shadowPosition.xy * (1.0 / shadowPosition.w), shadowPosition.z);
-    return ShadowSample_VSM(shadowMap, layer, position);
+    if (frameUniforms.shadowSamplingType == SHADOW_SAMPLING_RUNTIME_VSM) {
+        return ShadowSample_VSM(shadowMap, layer, shadowPosition);
+    }
```
Comment (romainguy) [inline shaders/src/shadowing.fs:380]: Not sure if it matters, but maybe use `else if` to help the compiler?

**Pull request merged**